internal/service/advancedclustertpf/resource.go

```diff
@@ -99,80 +96,230 @@ func (r *rs) Update(ctx context.Context, req resource.UpdateRequest, resp *resou
+	api := r.Client.AtlasV220240805.ClustersApi
+	projectID := tfModel.ProjectID.ValueString()
+	clusterName := tfModel.Name.ValueString()
+	_, _, err := api.UpdateCluster(ctx, projectID, clusterName, patchReq).Execute()
 	if err != nil {
-		diags.AddError("errorCreateAdvConfig", fmt.Sprintf(errorCreate, err.Error()))
-		return nil, true
+		diags.AddError("errorUpdate", fmt.Sprintf(errorUpdate, clusterName, err.Error()))
+		return nil
+	}
+	cluster, summary, detail := r.awaitChanges(ctx, &tfModel.Timeouts, diags, projectID, clusterName, "update")
+	if summary != "" || detail != "" {
+		diags.AddError(summary, detail)
+		return nil
 	}
-	sdkAdvConfigLegacy, err := ReadClusterProcessArgsResponseLegacy()
+	return cluster
+}
+
+func (r *rs) awaitChanges(ctx context.Context, t *timeouts.Value, diags *diag.Diagnostics, projectID, clusterName, changeReason string) (cluster *admin20240805.ClusterDescription20240805, summary, detail string) {
+	var timeoutDuration time.Duration
+	var localDiags diag.Diagnostics
+	var targetState = retrystrategy.RetryStrategyIdleState
+	var extraPending = []string{}
+	switch changeReason {
+	case "create":
+		timeoutDuration, localDiags = t.Create(ctx, defaultTimeout)
+		diags.Append(localDiags...)
+	case "update":
+		timeoutDuration, localDiags = t.Update(ctx, defaultTimeout)
+		diags.Append(localDiags...)
+	case "delete":
+		timeoutDuration, localDiags = t.Delete(ctx, defaultTimeout)
+		diags.Append(localDiags...)
+		targetState = retrystrategy.RetryStrategyDeletedState
+		extraPending = append(extraPending, retrystrategy.RetryStrategyIdleState)
+	default:
+		return nil, "errorReadingTimeout", "unknown change reason" + changeReason
+	}
+	if diags.HasError() {
+		return nil, "errorReadingTimeout", ""
+	}
+	stateConf := CreateStateChangeConfig(ctx, r.Client.AtlasV220240805, projectID, clusterName, targetState, timeoutDuration, extraPending...)
+	clusterAny, err := stateConf.WaitForStateContext(ctx)
 	if err != nil {
-		diags.AddError("errorCreateAdvConfigLegacy", fmt.Sprintf(errorCreate, err.Error()))
-		return nil, true
+		if admin.IsErrorCode(err, ErrorCodeClusterNotFound) && changeReason == "delete" {
+			return nil, "", ""
+		}
+		return nil, "errorAwaitingCluster", fmt.Sprintf(errorCreate, err)
+	}
+	if targetState == retrystrategy.RetryStrategyDeletedState {
+		return nil, "", ""
+	}
+	cluster, ok := clusterAny.(*admin20240805.ClusterDescription20240805)
+	if !ok {
+		return nil, "errorAwaitingCluster", fmt.Sprintf(errorCreate, "unexpected type from WaitForStateContext")
+	}
+	return cluster, "", ""
+}
+
+func (r *rs) convertClusterAddAdvConfig(ctx context.Context, legacyAdvConfig *admin20240530.ClusterDescriptionProcessArgs, advConfig *admin20240805.ClusterDescriptionProcessArgs20240805, cluster *admin20240805.ClusterDescription20240805, resourceTimeouts timeouts.Value, diags *diag.Diagnostics, state *tfsdk.State) (summary, detail string) {
```

Comment:
does convertClusterAddAdvConfig do more than doing the advanced config call? why does it have "convertCluster" in the name?

Reply:
Hopefully more clear now after refactoring